kafka-cron/consumer/main.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+	"bytes"
+	"log"
+	"os"
+	"os/exec"
+	"os/signal"
+	"sync"
+
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+
+	"encoding/json"
+
+	"github.com/Shopify/sarama"
+	"github.com/berkeli/kafka-cron/types"
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	brokerList = kingpin.Flag("brokerList", "List of brokers to connect").Default("localhost:9092").Strings()
+	topic      = kingpin.Flag("topic", "Topic name").Default("jobs-cluster-a").String()
+	retryTopic = kingpin.Flag("retryTopic", "Retry topic name").Default("jobs-cluster-a-retries").String()
+	partition  = kingpin.Flag("partition", "Partition number").Default("0").Int32()
+)
+
+func main() {
+
+	InitMonitoring(2112)
+
+	kingpin.Parse()
+	config := sarama.NewConfig()
+	config.Consumer.Return.Errors = true
+	brokers := *brokerList
+	master, err := sarama.NewConsumer(brokers, config)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	config = sarama.NewConfig()
+	config.Producer.RequiredAcks = sarama.WaitForAll
+	config.Producer.Return.Successes = true
+	producer, err := sarama.NewSyncProducer(*brokerList, config)
+	if err != nil {
+		log.Panic(err)
+	}
+	defer func() {
+		if err := producer.Close(); err != nil {
+			log.Panic(err)
+		}
+	}()
+
+	defer func() {
+		if err := master.Close(); err != nil {
+			log.Panic(err)
+		}
+	}()
+	consumer, err := master.ConsumePartition(*topic, *partition, sarama.OffsetNewest)
+	if err != nil {
+		log.Panic(err)
+	}
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+	chDone := make(chan bool)
+	wgWorkers := sync.WaitGroup{}
+	go func() {
+		for {
+			select {
+			case err := <-consumer.Errors():
+				log.Println(err)
+			case msg := <-consumer.Messages():
+				var cmd types.Command
+				err := json.Unmarshal(msg.Value, &cmd)
+				if err != nil {
+					log.Println(err)
+				}
+				wgWorkers.Add(1)
+				// TODO: Add a worker pool with semaphore? What if jobs are dependent on each other?
```

Comment:
there's no affordance in the system for the moment for job dependencies so i wouldn't worry about that

Reply:
some concept of a maximum running number of concurrent tasks is probably useful though